[Windows|Arm64|VarArgs] Correctly pass HFA arguments

jashook opened the pull request:
Windows arm64 VarArgs ABI requires passing using the general purpose registers. For HFA types, this would lead to bad codegen for vararg methods. Both the fixed arguments and the variable arguments are affected.

For arguments only, HFA structs need to be treated as if they are not HFAs.

CarolEidt commented on src/jit/compiler.cpp:
```diff
@@ -634,8 +634,14 @@ var_types Compiler::getPrimitiveTypeForStruct(unsigned structSize, CORINFO_CLASS
+#if !defined(_TARGET_UNIX_) && defined(_TARGET_ARM64_)
+                && !isVarArg // Arm64 Windows VarArg methods arguments will not
+                             // classify HFA types, they will need to be treated
+// as if they are not HFA types.
```
It would be nice to fix up this comment, if you can get the formatter to go along with a fix. Maybe put the comment above the `&&` line.

CarolEidt commented on src/jit/compiler.cpp:
```diff
@@ -847,8 +861,13 @@ var_types Compiler::getArgTypeForStruct(CORINFO_CLASS_HANDLE clsHnd,
-                // Structs that are pointer sized or smaller should have been handled by getPrimitiveTypeForStruct
-                assert(structSize > TARGET_POINTER_SIZE);
+#if !defined(_TARGET_UNIX_)
+                if (!isVarArg)
```
I don't see why this is needed. It should still be the case that `getPrimitiveTypeForStruct` should return a primitive type for anything <= `TARGET_POINTER_SIZE`, right?

jashook replied:
You are correct, thank you for catching this I do not know what I was thinking at the time.

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -2093,12 +2093,14 @@ void fgArgInfo::Dump(Compiler* compiler)
 //
 // Arguments:
-//    tmpVarNum  - the var num which we clone into the newly created temp var.
+//    bool isVarArg - is a VarArg method. Will effect HFA decision for arm64 windows
```
Nit: should be "affect"

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -1298,7 +1298,7 @@ class fgArgInfo
 
     void SortArgs();
 
-    void EvalArgsToTemps();
```
>EvalArgs [](start = 9, length = 8)

Could you change this to be in the positive sense? Perhaps `allowHfa`?

jashook replied:
Addressed

CarolEidt commented:
This looks good except for some minor things and one more substantive question.
I hope that eventually we can refactor both the incoming and outgoing arg handling so that we don't have to pass the hfa-related bools around all over the place.

jashook commented:
@CarolEidt @briansull can you take another look, I have updated and rebased against master. This pulls into #18358. I have also followed Carol's approach to using fgArgEntry to store isVararg.

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -2094,7 +2097,7 @@ void fgArgInfo::Dump(Compiler* compiler)
 //
 // Arguments:
-//    tmpVarNum  - the var num which we clone into the newly created temp var.
+//    curArgTabEntry
```
Thanks for catching that!

briansull commented on src/jit/compiler.cpp:
```diff
@@ -640,8 +640,15 @@ var_types Compiler::getPrimitiveTypeForStruct(unsigned structSize, CORINFO_CLASS
-            if (IsHfa(clsHnd))
+#else // !ARM_SOFTFP
+            if (IsHfa(clsHnd)
+#if !defined(_TARGET_UNIX_) && defined(_TARGET_ARM64_)
```
It might be better to define
```
#if defined(_TARGET_ARM64_) && !defined(_TARGET_UNIX_)
#define IGNORE_HFA_FOR_VARARG
```
Don't we have a TARGET_WINDOWS or WIN64?

jashook replied:
I believe there is only a WINOWS_AMD64_ABI define.

jashook replied:
I may be missing something.

jashook commented on src/jit/codegenarmarch.cpp:
```diff
@@ -2654,7 +2654,7 @@ void CodeGen::genJmpMethod(GenTree* jmp)
             gcInfo.gcMarkRegPtrVal(argReg, loadType);
 
-            if (compiler->lvaIsMultiregStruct(varDsc))
+            if (compiler->lvaIsMultiregStruct(varDsc, compiler->info.compIsVarArgs))
```
It will the decision for whether it is a multiregstruct for hfa cases. I believe it is necessary.

briansull replied:
sorry i was wrong

jashook commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

jashook commented:
osx timed out. Arm64 has passed. OSX has passed in last change, only difference are ifdef names. Which the build passes so I do not think it is required to wait.
